### ISSUE TITLE
uv_buf_init() truncates large buffers

### ISSUE THREAD

**Issue by trevnorris** [libuv/libuv#1233]
The problem API signature is `uv_buf_init(char*, unsigned int)`, causing an issue when called with large allocations (truncating the length to `0x80000000`).

Example:
```c
size_t len = 1024 * 1024 * 1024 * 6;  // 6GB
char* data = (char*)malloc(len);
assert(data != NULL);
uv_buf_t buf = uv_buf_init(data, len);
assert(buf.len == len);  // FAIL
```

I'd suggest the API should be changed to `uv_buf_init(char*, size_t)`. Even though this is not ABI compatible, I do hope it can make it into a minor release. If this is acceptable then I'd be happy to create the PR.

**Comment by kangjoni76** [comment]
shouldnt be ````uv_read_start```` could be used at once and emit chunk of buffers and parse byte per byte at ready_read function?

**Comment by bnoordhuis** [comment]
> Even though this is not ABI compatible

If it were an ABI break, it wouldn't be eligible.  However, I can't think of a supported system where it would actually be a change.  Hypothetically: a platform where `sizeof(int) != sizeof(size_t)` and:

1. Where integers of different sizes are passed in different registers, or

2. Where arguments are passed on the stack and int and size_t take up a different number of stack slots.

Either seems unlikely but that said, why do you need such a big buffer in the first place?  > 4 GB seems unreasonably large.

**Comment by trevnorris** [comment]
@bnoordhuis I was bit by this today while writing a PR for node's `fs.read()` to accept an `ArrayBuffer`. Since they can be arbitrarily large I was testing reading 6GB from disk. So given the following change:
```diff
    buf = buffer_data + off;
  
+   printf("len: %lu\n", len);              // prints 6442450944
    uv_buf_t uvbuf = uv_buf_init(const_cast<char*>(buf), len);
+   printf("uvbuf.len: %lu\n", uvbuf.len);  // prints 2147483648
  
    req = args[5];
  
```
This is confusing because the max `unsigned int` should be `4294967295`, but regardless the value is different than expected.

As for why it's necessary, personally I'm using it to process large data sets. There are several specific reasons why it's helpful, but I think saying it is helpful is sufficient. The Dat project has also [said it would be useful](https://twitter.com/trevnorris/status/809839922878132224) to be able to read/write more than 2GB of data at a time.

In addition, it's not intuitive to have `uv_buf_t`'s `len` property be a `size_t` but the argument to `uv_buf_init()` be an `unsigned int`.

The issue can be worked around by simply setting `uvbuf.len` manually, but the need to do so makes it feel like a flaw in the API.

**Comment by bnoordhuis** [comment]
> Where integers of different sizes are passed in different registers

I realized that's essentially what happens with the x86_64 SysV ABI.  With a function like `void f(int x)`, a caller is within its rights to pass `x` in %edi without having to care what's in the top-half of %rdi because the callee takes care of that.

Register moves zero-extend so that won't happen frequently but an optimizing compiler could decide that hey, that 32 bits value is already in the right place so I don't have to do anything but call the function, never mind that the top-half isn't zero.

In other words, changing the size argument isn't backwards compatible.  It can't land in v1.x but it's okay for v2.x.

**Comment by saghul** [comment]
Thanks for the detailed analysis, bet. Agreed!

@trevnorris While I agree this is a problem, why is it surfacing in Node? We never promised to fill the buffer up, so the caller would need to call the function again, with an offset. Or is it about some internal bookkeeping?

**Comment by bnoordhuis** [comment]
> Thanks for the detailed analysis, _bet_.

Autocomplete at work? =)

**Comment by saghul** [comment]
> Autocomplete at work? =)

I'm afraid so!

**Comment by trevnorris** [comment]
@saghul I'm making an addition to the `fs.write()`/`fs.read()` APIs to accept `ArrayBuffer`s. So the following will be valid:
```js
// Typed Arrays in V8 have a max 0x7fffffff indices, so this
// is the maximum size Uint32Array
const ab = new ArrayBuffer(0x1fffffffc);  // 8GB
// "file_path" is a file that contains a massive data matrix
const fd = fs.readSync(file_path, 'r+');
// Read in the entire file
fs.readSync(fd, ab, 0, ab.byteLength, 0);

// ... work with the data ...

// data matrix has changed, so write it to disk
fs.writeSync(fd, ab, 0, ab.byteLength, 0);
```

>  We never promised to fill the buffer up, so the caller would need to call the function again, with an offset.

I'm not sure what what you mean by "fill the buffer up". Can you clarify?

For large buffers the call already needs to be called more than once because the of kernel limitations on the size of a file that can be read/written, but I'd like this to be handled by the API. Regardless, I'd like to be able to pass in the entire buffer on the first attempt.

**Comment by indutny** [comment]
Doesn't `_LARGEFILE_SOURCE` imply 64-bit `size_t`? `int` is on other hand 32-bit, obviously...

**Comment by indutny** [comment]
Anyway, just for the sake of semantics we should be using `size_t`. That what this type is for, right?

**Comment by saghul** [comment]
@trevnorris Ah, I see what you mean now. While Node uses v1.x (which will for a while!) a workaround is needed: if the buffer size if > than UINT_MAX it should be capped and multiple (uv) read requests would be needed. I agree is not pretty, but we can't break the ABI.

**Comment by trevnorris** [comment]
@saghul isn't it enough to simply set `.len` manually? The field is already a `size_t`.

**Comment by saghul** [comment]
@trevnorris I meant when initializing the `uv_buf_t` buffer, so it would me obvious what the code is doing and why.

**Comment by moteus** [comment]
Maybe just add to node function like?

```C
uv_buf_t node_uv_buf_init(char* base, size_t len) {
  uv_buf_t buf;
  buf.base = base;
  buf.len = len;
  return buf;
}
```

**Comment by indutny** [comment]
Not that I mind it much, but this doesn't sound right.

**Comment by bnoordhuis** [comment]
> Mind sharing how you know that?

@trevnorris From reading the SysV ABI spec and LLVM's source code, verifying with a quick test program, and from sleeping on it for a night.  :-)

**Comment by saghul** [comment]
> Not that I mind it much, but this doesn't sound right.

@indutny what do you mean by "this" here?

**Comment by moteus** [comment]
On my test with MSVC2015 I got assert fail for both assertionst on x64 (as expected)
```C
  ULONG l = SIZE_MAX; // 4 bytes
  size_t s = SIZE_MAX; // 4 bytes on x86, 8 bytes on x64
  assert(l == s);
  assert(sizeof(l) >= sizeof(s));
```
Defenition of `WSABUF` form `ws2def.h` is
```C
typedef struct _WSABUF {
    ULONG len;     /* the length of the buffer */
    _Field_size_bytes_(len) CHAR FAR *buf; /* the pointer to the buffer */
} WSABUF, FAR * LPWSABUF;
```

**Comment by indutny** [comment]
Oooh... I see now. Sorry for the noise.
